Missing `log_prob` method or endpoint

Given a `model_name` and parameter values, call the `log_prob` function in the C++ model. Implementation can mirror that of the `param_names` endpoint.

When this is implemented, add a method to the PyStan 3 `Model` class.

(I think I have used this function one time in the last 4 years.)

I've exposed `log_prob` in HTTPStan and tested by hand. I am now writing a test, but I am met with the error below for **any** test that relies on `get_model_name`, any ideas what might cause this?

`============================================================================= test session starts ==============================================================================
platform linux -- Python 3.7.6, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /home/sgmcart3/Code/stan/httpstan/tests
plugins: asyncio-0.14.0
collected 1 item                                                                                                                                                               

test_linear_regression.py F

=================================================================================== FAILURES ===================================================================================
____________________________________________________________________________ test_linear_regression ____________________________________________________________________________

api_url = 'http://127.0.0.1:8080/v1'

    @pytest.mark.asyncio
    async def test_linear_regression(api_url: str) -> None:
        """Test sampling from linear regression posterior with defaults."""
    
        payload = {
            "function": "stan::services::sample::hmc_nuts_diag_e_adapt",
            "data": data,
            "num_samples": 500,
            "num_warmup": 500,
            "random_seed": 1,
        }
>       beta_0 = await helpers.sample_then_extract(api_url, program_code, payload, "beta.1")

test_linear_regression.py:46: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
helpers.py:146: in sample_then_extract
    operation = await sample(api_url, program_code, fit_payload)
helpers.py:84: in sample
    model_name = await get_model_name(api_url, program_code)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

api_url = 'http://127.0.0.1:8080/v1'
program_code = '\n    data {\n        int<lower=0> N;\n        int<lower=0> p;\n        matrix[N,p] x;\n        vector[N] y;\n    }\n...      vector[p] beta;\n        real<lower=0> sigma;\n    }\n    model {\n        y ~ normal(x * beta, sigma);\n    }\n'

    async def get_model_name(api_url: str, program_code: str) -> str:
        """Compile and retrieve model name.
    
        This function is a coroutine.
        """
        models_url = f"{api_url}/models"
        payload = {"program_code": program_code}
        async with aiohttp.ClientSession() as session:
            async with session.post(models_url, json=payload) as resp:
>               assert resp.status == 201
E               AssertionError

helpers.py:20: AssertionError
------------------------------------------------------------------------------ Captured log call -------------------------------------------------------------------------------
CRITICAL httpstan:views.py:119 Exception while building model extension module: `LinkError(DistutilsExecError("command 'g++' failed with exit status 1"))`, traceback: `['  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/httpstan/views.py", line 113, in handle_create_model\n    compiler_output = await httpstan.models.build_services_extension_module(program_code)\n', '  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/httpstan/models.py", line 176, in build_services_extension_module\n    None, httpstan.build_ext.run_build_ext, extensions, build_lib\n', '  File "/home/sgmcart3/anaconda3/lib/python3.7/concurrent/futures/thread.py", line 57, in run\n    result = self.fn(*self.args, **self.kwargs)\n', '  File "/home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/httpstan/build_ext.py", line 80, in run_build_ext\n    build_extension.run()\n', '  File "/home/sgmcart3/anaconda3/lib/python3.7/distutils/command/build_ext.py", line 340, in run\n    self.build_extensions()\n', '  File "/home/sgmcart3/anaconda3/lib/python3.7/distutils/command/build_ext.py", line 449, in build_extensions\n    self._build_extensions_serial()\n', '  File "/home/sgmcart3/anaconda3/lib/python3.7/distutils/command/build_ext.py", line 474, in _build_extensions_serial\n    self.build_extension(ext)\n', '  File "/home/sgmcart3/anaconda3/lib/python3.7/distutils/command/build_ext.py", line 559, in build_extension\n    target_lang=language)\n', '  File "/home/sgmcart3/anaconda3/lib/python3.7/distutils/ccompiler.py", line 717, in link_shared_object\n    extra_preargs, extra_postargs, build_temp, target_lang)\n', '  File "/home/sgmcart3/anaconda3/lib/python3.7/distutils/unixccompiler.py", line 207, in link\n    raise LinkError(msg)\n']`
=============================================================================== warnings summary ===============================================================================
../../httpstan-development/lib/python3.7/site-packages/setuptools/depends.py:2
  /home/sgmcart3/Code/stan/httpstan-development/lib/python3.7/site-packages/setuptools/depends.py:2: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================================================================== short test summary info ============================================================================
FAILED test_linear_regression.py::test_linear_regression - AssertionError
======================================================================== 1 failed, 1 warning in 25.22s =========================================================================`

--- a/httpstan/openapi.py
+++ b/httpstan/openapi.py
@@ -42,6 +42,7 @@ def openapi_spec() -> apispec.APISpec:
     spec.path(path="/v1/models", view=views.handle_list_models)
     spec.path(path="/v1/models/{model_id}", view=views.handle_delete_model)
     spec.path(path="/v1/models/{model_id}/params", view=views.handle_show_params)
+    spec.path(path="/v1/models/{model_id}/log_prob", view=views.handle_log_prob)
     spec.path(path="/v1/models/{model_id}/fits", view=views.handle_create_fit)
     spec.path(path="/v1/models/{model_id}/fits/{fit_id}", view=views.handle_get_fit)
     spec.path(path="/v1/models/{model_id}/fits/{fit_id}", view=views.handle_delete_fit)

--- a/httpstan/routes.py
+++ b/httpstan/routes.py
@@ -20,6 +20,7 @@ def setup_routes(app: aiohttp.web.Application) -> None:
     app.router.add_get("/v1/models", views.handle_list_models)
     app.router.add_delete("/v1/models/{model_id}", views.handle_delete_model)
     app.router.add_post("/v1/models/{model_id}/params", views.handle_show_params)
+    app.router.add_post("/v1/models/{model_id}/log_prob", views.handle_log_prob)
     app.router.add_post("/v1/models/{model_id}/fits", views.handle_create_fit)
     app.router.add_get("/v1/models/{model_id}/fits/{fit_id}", views.handle_get_fit)
     app.router.add_delete("/v1/models/{model_id}/fits/{fit_id}", views.handle_delete_fit)

--- a/httpstan/schemas.py
+++ b/httpstan/schemas.py
@@ -181,3 +181,11 @@ class WriterMessage(marshmallow.Schema):
     topic = fields.String(required=True, validate=validate.OneOf(["logger", "initialization", "sample", "diagnostic"]))
     # values is either a List or a Mapping. Marshmallow lacks a union type.
     values = fields.Raw(required=True)
+
+
+class ShowLogProbRequest(marshmallow.Schema):
+    """Schema for log_prob request."""
+
+    data = fields.Nested(Data(), missing={})
+    unconstrained_parameters = fields.List(fields.Float(), required=True)
+    adjust_transform = fields.Boolean(missing=True)

--- a/httpstan/stan_services.cpp
+++ b/httpstan/stan_services.cpp
@@ -122,6 +122,34 @@ std::vector<std::vector<size_t>> dims(py::dict data) {
 }
 
 // See exported docstring
+double log_prob(py::dict data,
+                const std::vector<double>& unconstrained_parameters,
+                bool adjust_transform) {
+  stan::io::array_var_context &var_context = new_array_var_context(data);
+  // random_seed, the second argument, is unused but the function requires it.
+  stan::model::model_base &model = new_model(var_context, (unsigned int)1, &std::cout);
+  if(unconstrained_parameters.size() != model.num_params_r()) {
+    throw std::runtime_error("The number of parameters does not match the number of unconstrained parameters in the model.");
+  }
+  // copy unconstrained_parameters -> params_r
+  std::vector<double> params_r(unconstrained_parameters.size());
+  std::memcpy(params_r.data(), unconstrained_parameters.data(), unconstrained_parameters.size()*sizeof(double));
+  // calculate logprob
+  std::vector<int> params_i(model.num_params_i(), 0);
+  double lp;
+  if(adjust_transform) {
+    lp = stan::model::log_prob_propto<true>(model, params_r, params_i, &std::cout);
+  } else {
+    lp = stan::model::log_prob_propto<false>(model, params_r, params_i, &std::cout);
+  }
+
+  delete &model;
+  delete &var_context;
+
+  return lp;
+}
+
+// See exported docstring
 int hmc_nuts_diag_e_adapt_wrapper(std::string socket_filename, py::dict data, py::dict init, int random_seed,
                                   int chain, double init_radius, int num_warmup, int num_samples, int num_thin,
                                   bool save_warmup, int refresh, double stepsize, double stepsize_jitter,
@@ -177,6 +205,8 @@ PYBIND11_MODULE(stan_services, m) {
   m.def("constrained_param_names", &constrained_param_names, py::arg("data"),
         "Call the ``constrained_param_names`` method of the model.");
   m.def("dims", &dims, py::arg("data"), "Call the ``get_dims`` method of the model.");
+  m.def("log_prob", &log_prob, py::arg("data"), py::arg("unconstrained_parameters"), py::arg("adjust_transform"),
+        "Call stan::model::log_prob_propto");
   m.def("hmc_nuts_diag_e_adapt_wrapper", &hmc_nuts_diag_e_adapt_wrapper, py::arg("socket_filename"), py::arg("data"),
         py::arg("init"), py::arg("random_seed"), py::arg("chain"), py::arg("init_radius"), py::arg("num_warmup"),
         py::arg("num_samples"), py::arg("num_thin"), py::arg("save_warmup"), py::arg("refresh"), py::arg("stepsize"),

--- a/httpstan/views.py
+++ b/httpstan/views.py
@@ -558,3 +558,79 @@ async def handle_get_operation(request: aiohttp.web.Request) -> aiohttp.web.Resp
         message, status = f"Operation `{operation_name}` not found.", 404
         return aiohttp.web.json_response(_make_error(message, status=status), status=status)
     return aiohttp.web.json_response(operation)
+
+
+async def handle_log_prob(request: aiohttp.web.Request) -> aiohttp.web.Response:
+    """Calculate the log probability.
+
+    ---
+    post:
+      summary: Return the log probability of the unconstrained parameters.
+      description: >-
+        Returns the output of Stan C++ `stan::model::log_prob_propto`.
+      consumes:
+        - application/json
+      produces:
+        - application/json
+      parameters:
+        - name: model_id
+          in: path
+          description: ID of Stan model to use
+          required: true
+          type: string
+        - in: body
+          name: data
+          description: >-
+              Data for the Stan Model.
+          required: true
+          schema: Data
+        - in: body
+          name: unconstrained_parameters
+          description: >-
+              Unconstrained parameters to calculate log probability for.
+          required: true
+          schema:
+            type: boolean
+        - in: body
+          name: adjust_transform
+          description: >-
+              Boolean to control whether we apply a Jacobian adjust transform.
+          required: true
+          schema:
+            type: array
+            items:
+              type: number
+      responses:
+        "200":
+          description: Log probability of the unconstrained parameters.
+          schema:
+            type: object
+            properties:
+              log_prob:
+                type: number
+        "400":
+          description: Error associated with request.
+          schema: Status
+        "404":
+          description: Model not found.
+          schema: Status
+    """
+    args = await webargs.aiohttpparser.parser.parse(schemas.ShowLogProbRequest(), request)
+    model_name = f'models/{request.match_info["model_id"]}'
+    data = args["data"]
+    unconstrained_parameters = args["unconstrained_parameters"]
+    adjust_transform = args["adjust_transform"]
+
+    try:
+        services_module = httpstan.models.import_services_extension_module(model_name)
+    except KeyError:
+        message, status = f"Model `{model_name}` not found.", 404
+        return aiohttp.web.json_response(_make_error(message, status=status), status=status)
+
+    try:
+        lp = services_module.log_prob(data, unconstrained_parameters, adjust_transform)  # type: ignore
+    except Exception as exc:
+        message, status = f"Error calling log_prob: `{exc}`", 400
+        logger.critical(message)
+        return aiohttp.web.json_response(_make_error(message, status=status), status=status)
+    return aiohttp.web.json_response({"log_prob": lp}, status=200)

--- a/tests/test_log_prob.py
+++ b/tests/test_log_prob.py
@@ -1,0 +1,73 @@
+"""Test log_prob endpoint through a Gaussian toy problem."""
+import random
+
+import aiohttp
+import numpy as np
+import pytest
+
+import helpers
+
+program_code = """
+parameters {
+  real y;
+}
+model {
+  y ~ normal(0, 1);
+}
+"""
+
+x = random.uniform(0, 10)
+
+
+def gaussian_lp(x: float, mean: float, var: float) -> float:
+    """Analytically evaluate Gaussian log probability."""
+
+    lp = -1 / (2 * var) * (x - mean) ** 2
+    return lp
+
+
+@pytest.mark.parametrize("x", [x])
+@pytest.mark.asyncio
+async def test_log_prob_analytically(x: float, api_url: str) -> None:
+    """Test log probability endpoint."""
+
+    model_name = await helpers.get_model_name(api_url, program_code)
+    models_params_url = f"{api_url}/{model_name}/log_prob"
+    payload = {"data": {}, "unconstrained_parameters": [x], "adjust_transform": False}
+    async with aiohttp.ClientSession() as session:
+        async with session.post(models_params_url, json=payload) as resp:
+            assert resp.status == 200
+            response_payload = await resp.json()
+            assert "log_prob" in response_payload
+            httpstan_lp = response_payload["log_prob"]
+            lp = gaussian_lp(x, 0, 1)
+            assert np.allclose(httpstan_lp, lp)
+
+
+@pytest.mark.asyncio
+async def test_log_prob_sampling(api_url: str) -> None:
+    """Test log probability endpoint against sampled model."""
+
+    payload = {
+        "function": "stan::services::sample::hmc_nuts_diag_e_adapt",
+        "data": {},
+        "num_samples": 500,
+        "num_warmup": 500,
+        "random_seed": 1,
+    }
+    operation = await helpers.sample(api_url, program_code, payload)
+    fit_name = operation["result"]["name"]
+    fit_bytes_ = await helpers.fit_bytes(api_url, fit_name)
+    assert isinstance(fit_bytes_, bytes)
+    sample_lp = helpers.extract("lp__", fit_bytes_)[0]
+    sample_y = helpers.extract("y", fit_bytes_)[0]
+    model_name = await helpers.get_model_name(api_url, program_code)
+    models_params_url = f"{api_url}/{model_name}/log_prob"
+    payload = {"data": {}, "unconstrained_parameters": [sample_y], "adjust_transform": False}
+    async with aiohttp.ClientSession() as session:
+        async with session.post(models_params_url, json=payload) as resp:
+            assert resp.status == 200
+            response_payload = await resp.json()
+            assert "log_prob" in response_payload
+            httpstan_lp = response_payload["log_prob"]
+            assert np.allclose(httpstan_lp, sample_lp)

